lib/mayaUsd/ufe/UsdStageMap.cpp

```diff
@@ -85,32 +110,53 @@ void UsdStageMap::addItem(const Ufe::Path& path, UsdStageWeakPtr stage)
         return;
     }
 
-    // Convert the tail of the UFE path to an MObjectHandle.
-    auto proxyShape = proxyShapeHandle(path);
+    // Convert the UFE path to an MObjectHandle.
+    auto proxyShape = nameLookup(path);
     if (!proxyShape.isValid()) {
         return;
     }
 
-    // Could get the stage from the proxy shape object in the stage() method,
-    // but since it's given here, simply store it.
-    fObjectToStage[proxyShape] = stage;
+    // We've just done a name-based lookup of the proxy shape, so the stage
+    // cannot be null.
+    auto stage = objToStage(proxyShape.object());
+    TF_AXIOM(stage);
+
+    fPathToObject[path] = proxyShape;
     fStageToObject[stage] = proxyShape;
 }
 
-UsdStageWeakPtr UsdStageMap::stage(const Ufe::Path& path)
+UsdStageWeakPtr UsdStageMap::stage(const Ufe::Path& path) { return objToStage(proxyShape(path)); }
+
+MObject UsdStageMap::proxyShape(const Ufe::Path& path)
```

Comment:
This is the engine method for the stage() and proxyShapeNode() methods.  It indexes into the unordered map of MObjectHandle's to proxy shapes using the Ufe::Path hash.  This is how we avoid name-based lookups.